lib/collection/src/shards/shard_holder/resharding.rs

```diff
@@ -344,51 +344,55 @@ impl ShardHolder {
         }
 
         // Remove new shard if resharding up
-        if direction == ReshardingDirection::Up {
+        if (is_in_progress || force) && direction == ReshardingDirection::Up {
             if let Some(shard) = self.get_shard(shard_id) {
-                match shard.peer_state(peer_id) {
-                    Some(ReplicaState::Resharding) => {
-                        log::debug!("removing peer {peer_id} from {shard_id} replica set");
-                        shard.remove_peer(peer_id).await?;
-                    }
-
-                    Some(ReplicaState::Dead) if is_in_progress => {
-                        log::debug!("removing dead peer {peer_id} from {shard_id} replica set");
-                        shard.remove_peer(peer_id).await?;
-                    }
-
-                    Some(state) => {
-                        return Err(CollectionError::bad_request(format!(
-                            "peer {peer_id} is in {state:?} state"
-                        )));
-                    }
-
-                    None => {
-                        log::warn!(
-                            "aborting resharding {resharding_key}, \
-                             but peer {peer_id} does not exist in {shard_id} replica set"
-                        );
+                // Ensure all replicas are in expected state
+                for (peer_id, replica_state) in shard.peers() {
+                    match replica_state {
+                        // Valid replica states:
+                        // - in resharding state during migration transfer
+                        // - in active state after migration transfer
+                        // - in partial state during replication
+                        // - in dead state during errors
+                        ReplicaState::Resharding
+                        | ReplicaState::Active
+                        | ReplicaState::Partial
+                        | ReplicaState::Dead => {}
+
```

Comment:
Question: maybe we want to allow any state if `force == true`?

Reply:
> I think, if the whole condition is under is_in_progress, we can allow removing replicas in any state.

Fine by me.

I figured this being a bit more strict will prevent problems if something else is interfering with the replicas (some user transfer potentially).